src/esp/sensor/OrthoCamera.cpp

```diff
@@ -0,0 +1,180 @@
+
+OrthoCamera& OrthoCamera::setViewport(gfx::RenderCamera& targetCamera) {
+  targetCamera.setViewport(this->framebufferSize());
+  return *this;
+}
+
+bool OrthoCamera::getObservationSpace(ObservationSpace& space) {
+  space.spaceType = ObservationSpaceType::TENSOR;
+  space.shape = {static_cast<size_t>(spec_->resolution[0]),
+                 static_cast<size_t>(spec_->resolution[1]),
+                 static_cast<size_t>(spec_->channels)};
+  space.dataType = core::DataType::DT_UINT8;
+  if (spec_->sensorType == SensorType::SEMANTIC) {
+    space.dataType = core::DataType::DT_UINT32;
+  } else if (spec_->sensorType == SensorType::DEPTH) {
+    space.dataType = core::DataType::DT_FLOAT;
+  }
+  return true;
+}
+
+bool OrthoCamera::getObservation(sim::Simulator& sim, Observation& obs) {
+  // TODO: check if sensor is valid?
+  // TODO: have different classes for the different types of sensors
+  //
+  if (!hasRenderTarget())
+    return false;
+
+  drawObservation(sim);
+  readObservation(obs);
+
+  return true;
+}
+
+bool OrthoCamera::drawObservation(sim::Simulator& sim) {
+  if (!hasRenderTarget()) {
+    return false;
+  }
+
+  renderTarget().renderEnter();
+
+  gfx::RenderCamera::Flags flags;
+  if (sim.isFrustumCullingEnabled())
+    flags |= gfx::RenderCamera::Flag::FrustumCulling;
+
+  gfx::Renderer::ptr renderer = sim.getRenderer();
+  if (spec_->sensorType == SensorType::SEMANTIC) {
+    // TODO: check sim has semantic scene graph
+    renderer->draw(*this, sim.getActiveSemanticSceneGraph(), flags);
+    if (&sim.getActiveSemanticSceneGraph() != &sim.getActiveSceneGraph()) {
+      flags |= gfx::RenderCamera::Flag::ObjectsOnly;
+      renderer->draw(*this, sim.getActiveSceneGraph(), flags);
+    }
+  } else {
+    // SensorType is DEPTH or any other type
+    renderer->draw(*this, sim.getActiveSceneGraph(), flags);
+  }
+
+  renderTarget().renderExit();
+
+  return true;
+}
+
+void OrthoCamera::readObservation(Observation& obs) {
+  // Make sure we have memory
+  if (buffer_ == nullptr) {
+    // TODO: check if our sensor was resized and resize our buffer if needed
+    ObservationSpace space;
+    getObservationSpace(space);
+    buffer_ = core::Buffer::create(space.shape, space.dataType);
+  }
+  obs.buffer = buffer_;
+
+  // TODO: have different classes for the different types of sensors
+  // TODO: do we need to flip axis?
+  if (spec_->sensorType == SensorType::SEMANTIC) {
+    renderTarget().readFrameObjectId(Magnum::MutableImageView2D{
+        Magnum::PixelFormat::R32UI, renderTarget().framebufferSize(),
+        obs.buffer->data});
+  } else if (spec_->sensorType == SensorType::DEPTH) {
+    renderTarget().readFrameDepth(Magnum::MutableImageView2D{
+        Magnum::PixelFormat::R32F, renderTarget().framebufferSize(),
+        obs.buffer->data});
+  } else {
+    renderTarget().readFrameRgba(Magnum::MutableImageView2D{
+        Magnum::PixelFormat::RGBA8Unorm, renderTarget().framebufferSize(),
+        obs.buffer->data});
+  }
```

Comment:
Future Proof by doing an elif notImplemented complaint.

Reply:
Or maybe a switch statement would be more readable here?

Reply:
This code was lifted verbatim from pinhole. Probably ought to be moved to VisualSensor.

Reply:
Ah leave it as then. @bigbike Something to think about during sensor redesign.